### PR TITLE
[Mypy] Part 3 fix typing for nested directories for most of directory

### PR DESCRIPTION
**Pull request by rkooo567**
This fixes typing for most of the nested directories. 

The remaining one has a lot of changes required for each of them. 

Also move --follow-imports to the pyproject.toml

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/distributed/device_communicators/pynccl.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ class ncclDataType_t(ctypes.c_int):
     ncclNumTypes = 10
 
     @classmethod
-    def from_torch(cls, dtype: torch.dtype) -> 'ncclDataType_t':
+    def from_torch(cls, dtype: torch.dtype) -> int:
```
Comment (simon-mo) [inline vllm/distributed/device_communicators/pynccl.py:145]: this is weird that mypy can't handle this, it should defenitely return the `c_int`

**Review by simon-mo — APPROVED** [review]
unsure about the nccl's ctypes part.

File: vllm/distributed/device_communicators/pynccl.py (hunk 1 above)
Comment (rkooo567) [reply vllm/distributed/device_communicators/pynccl.py:145]: Hmm when I added a breakpoint, it does return int. 

```
(Pdb) ncclDataType_t.from_torch(tensor.dtype)
7
(Pdb) type(ncclDataType_t.from_torch(tensor.dtype))
<class 'int'>
```

Do you think it is a bug?

Comment (youkaichao) [reply vllm/distributed/device_communicators/pynccl.py:145]: Well, technically, this function indeed returns `int`. `ctypes` will automatically convert int to `c_int` . Note that `ctypes.c_int` is a class/container to hold an `int`, and itself is not an `int`.

TL;DR; `def from_torch(cls, dtype: torch.dtype) -> int:` is correct.

Comment (rkooo567) [reply vllm/distributed/device_communicators/pynccl.py]:
```
# enums
class ncclDataType_t(ctypes.c_int):
    pass


class ncclDataTypeEnum:
    ncclInt8 = ncclDataType_t(0)
    ncclChar = ncclDataType_t(0)
    ncclUint8 = ncclDataType_t(1)
    ncclInt32 = ncclDataType_t(2)
    ncclInt = ncclDataType_t(2)
    ncclUint32 = ncclDataType_t(3)
    ncclInt64 = ncclDataType_t(4)
    ncclUint64 = ncclDataType_t(5)
    ncclFloat16 = ncclDataType_t(6)
    ncclHalf = ncclDataType_t(6)
    ncclFloat32 = ncclDataType_t(7)
    ncclFloat = ncclDataType_t(7)
    ncclFloat64 = ncclDataType_t(8)
    ncclDouble = ncclDataType_t(8)
    ncclBfloat16 = ncclDataType_t(9)
    ncclNumTypes = ncclDataType_t(10)

    @classmethod
    def from_torch(cls, dtype: torch.dtype) -> "ncclDataType_t":
        if dtype == torch.int8:
            return cls.ncclInt8
        if dtype == torch.uint8:
            return cls.ncclUint8
        if dtype == torch.int32:
            return cls.ncclInt32
        if dtype == torch.int64:
            return cls.ncclInt64
        if dtype == torch.float16:
            return cls.ncclFloat16
        if dtype == torch.float32:
            return cls.ncclFloat32
        if dtype == torch.float64:
            return cls.ncclFloat64
        if dtype == torch.bfloat16:
            return cls.ncclBfloat16
        raise ValueError(f"Unsupported dtype: {dtype}")
```

I fixed this way so that it returns the correct type.

Comment (rkooo567) [reply vllm/distributed/device_communicators/pynccl.py]: oh @youkaichao I just saw your msg. So do you think ^ makes more sense or just return int? I can follow your suggestion

Comment (rkooo567) [reply vllm/distributed/device_communicators/pynccl.py]: I think technically inheriting c_int here is not necessary then? it is just like a simple enum

Comment (youkaichao) [reply vllm/distributed/device_communicators/pynccl.py]: yes, that's ChatGPT's fault 😉

Comment (rkooo567) [reply vllm/distributed/device_communicators/pynccl.py:130]: Fixed. PTAL...https://github.com/vllm-project/vllm/pull/4161/commits/81313dc5a1d32bc3cc2aed2a93304c6cfffedf39

File: vllm/distributed/device_communicators/pynccl.py [hunk 2]
```diff
@@ -180,8 +179,8 @@ def from_torch(cls, op: ReduceOp) -> 'ncclRedOp_t':
 _c_ncclAllReduce = nccl.ncclAllReduce
 _c_ncclAllReduce.restype = ctypes.c_int
 _c_ncclAllReduce.argtypes = [
-    ctypes.c_void_p, ctypes.c_void_p, ctypes.c_size_t, ncclDataType_t,
-    ncclRedOp_t, ctypes.c_void_p, ctypes.c_void_p
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/pynccl.py:184]: This is not intuitive. Let's keep `ncclDataType_t` and `ncclRedOp_t` here.

Does mypy understand the call signature of `_c_ncclAllReduce`? I suppose it will just ignore them. And we can leave a comment here, saying that `int` will be automatically converted by `ctypes`.

File: vllm/distributed/device_communicators/pynccl.py [hunk 3]
```diff
@@ -180,8 +179,8 @@ def from_torch(cls, op: ReduceOp) -> 'ncclRedOp_t':
...
+    ctypes.c_void_p, ctypes.c_void_p, ctypes.c_size_t, ctypes.c_int,
+    ctypes.c_void_p, ctypes.c_void_p
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/pynccl.py:183]: This is not intuitive. Let's keep `ncclDataType_t` and `ncclRedOp_t` here.

Does mypy understand the call signature of `_c_ncclAllReduce`? I suppose it will just ignore them. And we can leave a comment here, saying that `int` will be automatically converted by `ctypes`.

Comment (rkooo567) [reply vllm/distributed/device_communicators/pynccl.py:183]: I found ncclDataType_t being a random enum and type at the same time a bit weird though. Why don't we then just

```
ncclDataType_t(ctype.c_int):
    pass

ncclDataEnum:
    ... = 0
    .... = 1
```
?

Comment (youkaichao) [reply vllm/distributed/device_communicators/pynccl.py:183]: Well, that's also doable.

Please use `ncclDataType_t  = ctype.c_int`, and `ncclDataTypeEnum`

Comment (rkooo567) [reply vllm/distributed/device_communicators/pynccl.py:183]: sgtm!

**Comment by rkooo567** [comment]
It is ready to merge!
